libraries/extensions/ros2-bridge/python/src/typed/serialize/mod.rs

```diff
@@ -0,0 +1,184 @@
+use std::{borrow::Cow, collections::HashMap, fmt::Display};
+
+use arrow::array::{Array, ArrayRef, AsArray};
+use dora_ros2_bridge_msg_gen::types::{
+    primitives::{GenericString, NestableType},
+    MemberType,
+};
+use serde::ser::SerializeStruct;
+
+use super::{TypeInfo, DUMMY_FIELD_NAME, DUMMY_STRUCT_NAME};
+
+mod array;
+mod defaults;
+mod primitive;
+mod sequence;
+
+#[derive(Debug, Clone)]
+pub struct TypedValue<'a> {
+    pub value: &'a ArrayRef,
+    pub type_info: &'a TypeInfo<'a>,
+}
+
+impl serde::Serialize for TypedValue<'_> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let empty = HashMap::new();
+        let package_messages = self
+            .type_info
+            .messages
+            .get(self.type_info.package_name.as_ref())
+            .unwrap_or(&empty);
+        let message = package_messages
+            .get(self.type_info.message_name.as_ref())
+            .ok_or_else(|| {
+                error(format!(
+                    "could not find message type {}::{}",
+                    self.type_info.package_name, self.type_info.message_name
+                ))
+            })?;
+
+        let input = self
+            .value
+            .as_struct_opt()
+            .ok_or_else(|| error("expected struct array"))?;
+        for column_name in input.column_names() {
+            if !message.members.iter().any(|m| m.name == column_name) {
+                return Err(error(format!(
+                    "given struct has unknown field {column_name}"
+                )))?;
+            }
+        }
+        if input.is_empty() {
+            // TODO: publish default value
+            return Err(error("given struct is empty"))?;
+        }
+        if input.len() > 1 {
+            return Err(error(format!(
+                "expected single struct instance, got struct array with {} entries",
+                input.len()
+            )))?;
+        }
+        let mut s = serializer.serialize_struct(DUMMY_STRUCT_NAME, message.members.len())?;
```

Comment:
It seems that the serializer use `serialize_struct` but the deserializer use `deserialize_tuple_struct`, and I was wondering if we can use the same one?

I was trying to build a roundtrip between the serializer and deserializer, and it seems that the token stream is not the same.

Reply:
I have made a quick branch where you can checkout my WIP at: https://github.com/dora-rs/dora/tree/test-to-ros2-fix
Thanks for this quick branch @phil-opp :)

Reply:
Thanks for the review!

> It seems that the serializer use `serialize_struct` but the deserializer use `deserialize_tuple_struct`, and I was wondering if we can use the same one?

Good catch! There should not be a difference between the two in the CDR deserialization format, but we should still use the same one to be consistent.

> I was trying to build a roundtrip between the serializer and deserializer, and it seems that the token stream is not the same.

That's a good idea! I looked into your branch and I discovered one error where we forgot to wrap sequences/arrays into a ListArray on deserialization. I will push a fix in a minute.

With that bug fixed, the roundtrip almost works, but the struct field order is different in some cases. This is kind of expected because we always use the field order specified in the ROS2 message file on serialization, but permit arbitrary field orders as input for serialization. However, even with the Python array specified in the exact same order, we get some reordering when using `pa.array`. I suspect that this is because of https://github.com/apache/arrow/issues/34250. This is independent from our ROS2 bridge though.